stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,454 @@
+  std::vector<double> deltac(dim);
+
+  for (std::size_t i = 0; i != dim; i++) {
+    c[i] = (a[i] + b[i]) / 2;
+  }
+
+  for (std::size_t i = 0; i != dim; i++) {
+    deltac[i] = fabs(b[i] - a[i]) / 2;
+  }
+  double v = 1.0;
+  for (std::size_t i = 0; i != dim; i++) {
+    v *= deltac[i];
+  }
+
+  if (v == 0.0) {
+    return std::make_tuple(0.0, 0.0, 0);
+  }
+
+  double f1 = integrand(c, pars);
+  double f2 = 0.0;
+  double f3 = 0.0;
+  double twelvef1 = 12 * f1;
+
+  double maxdivdiff = 0.0;
+  std::vector<double> divdiff(dim);
+  std::vector<double> p2(dim);
+  std::vector<double> p3(dim);
+  std::vector<double> cc(dim, 0);
+
+  for (std::size_t i = 0; i != dim; i++) {
+    for (std::size_t j = 0; j != dim; j++) {
+      p2[j] = deltac[j] * p[0][i][j];
+    }
+
+    for (std::size_t j = 0; j != dim; j++) {
+      cc[j] = c[j] + p2[j];
+    }
+    double f2i = integrand(cc, pars);
+    for (std::size_t j = 0; j != dim; j++) {
+      cc[j] = c[j] - p2[j];
+    }
+    double temp = integrand(cc, pars);
+    f2i += temp;
+
+    for (std::size_t j = 0; j != dim; j++) {
+      p3[j] = deltac[j] * p[1][i][j];
+    }
+    for (std::size_t j = 0; j != dim; j++) {
+      cc[j] = c[j] + p3[j];
+    }
+    double f3i = integrand(cc, pars);
+    for (std::size_t j = 0; j != dim; j++) {
+      cc[j] = c[j] - p3[j];
+    }
+    temp = integrand(cc, pars);
+    f3i += temp;
+    f2 += f2i;
+    f3 += f3i;
+    divdiff[i] = fabs(f3i + twelvef1 - 7 * f2i);
+  }
+  std::vector<double> p4(dim);
+  double f4 = 0.0;
+  for (std::size_t i = 0; i != p[2].size(); i++) {
+    for (std::size_t j = 0; j != dim; j++) {
+      p4[j] = deltac[j] * p[2][i][j];
+    }
+    for (std::size_t j = 0; j != dim; j++) {
+      cc[j] = c[j] + p4[j];
+    }
+    double temp = integrand(cc, pars);
+    f4 += temp;
+  }
+  double f5 = 0.0;
+  std::vector<double> p5(dim);
+  for (std::size_t i = 0; i != p[3].size(); i++) {
+    for (std::size_t j = 0; j != dim; j++) {
+      p5[j] = deltac[j] * p[3][i][j];
+    }
+
+    for (std::size_t j = 0; j != dim; j++) {
+      cc[j] = c[j] + p5[j];
+    }
+    double temp = integrand(cc, pars);
+    f5 += temp;
+  }
+
+  double I = v * (w[0] * f1 + w[1] * f2 + w[2] * f3 + w[3] * f4 + w[4] * f5);
+  double Idash = v * (wd[0] * f1 + wd[1] * f2 + wd[2] * f3 + wd[3] * f4);
+  double E = fabs(I - Idash);
+
+  int kdivide = 0;
+  double deltaf = E / (std::pow(10, dim) * v);
+  for (std::size_t i = 0; i != dim; i++) {
+    double delta = divdiff[i] - maxdivdiff;
+    if (delta > deltaf) {
+      kdivide = i;
+      maxdivdiff = divdiff[i];
+    } else if ((fabs(delta) <= deltaf) && (deltac[i] > deltac[kdivide])) {
+      kdivide = i;
+    }
+  }
+  return std::make_tuple(I, E, kdivide);
+}
+
+struct Box {
+  Box(const std::vector<double>& a, const std::vector<double>& b, double I,
+      double err, int kdivide)
+      : a(a), b(b), I(I), E(err), kdiv(kdivide) {}
+  bool operator>(const Box& box) const { return E < box.E; }
```

Comment:
```suggestion
  bool operator<(const Box& box) const { return E < box.E; }
```
I think it's more logical to not reverse the ordering.

Reply:
Ah, ok, I got it. Yes, true, I changed the symbol.
Next commit.